marconi/src/Marconi/IndexersHotStore.hs

```diff
@@ -0,0 +1,172 @@
+{-# LANGUAGE FlexibleContexts     #-}
+{-# LANGUAGE GADTs                #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{- | IndexerHotstore provides in concurrent in memory cache for marcoin clients
+-}
+module Marconi.IndexersHotStore(
+    puts
+    , findByAddress
+    , bootstrapHotStore
+    , bootstrapHotStoreWithKeys
+    , IndexerHotStore
+    , targetAddressParser
+    , TargetAddresses
+    , keys
+    , isInTargetTxOut
+    , findAll
+    ,isTargetTxOut
+    ) where
+
+import Cardano.Api qualified as CApi
+import Control.Concurrent.STM.TVar (TVar, modifyTVar, newTVarIO, readTVarIO)
+import Control.Monad.STM (atomically)
+import Data.Foldable (traverse_)
+import Data.List.NonEmpty (NonEmpty, fromList, nub)
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map (adjust, empty, fromList, keysSet, lookup, toList)
+import Data.Proxy (Proxy (Proxy))
+import Data.Set (Set)
+import Data.Set qualified as Set (empty, insert)
+import Data.Text (Text, pack, unpack)
+import Ledger.Address (Address)
+import Ledger.Tx (TxOutRef)
+import Ledger.Tx.CardanoAPI (ToCardanoError (DeserialisationError, Tag), fromCardanoAddress)
+import Marconi.Index.Utxo qualified as Utxo
+
+-- | Key value store where keys point to set of unique values
+-- TODO consider using unordered container
+newtype KeyValues k v = KeyValues  {unKeyValues :: Map k (Set v) }
+
+-- | Key value hotstore
+newtype HotStore k v = HotStore { unHot :: TVar (KeyValues k v) }
+
+-- | Key value hotstore implementation for Plutus Address and Set of TxOutRefs
+type IndexerHotStore = HotStore Address  TxOutRef
+
+-- | Allows clients to store a list of TxOutRefs keyed by Plutus address
+puts :: IndexerHotStore -- ^ hot store
+    -> [Utxo.UtxoRow]  -- ^ UtxoRows to insert into hotstore
+    -> IO ()
+puts hotStore = traverse_ (put hotStore)
+
+put :: IndexerHotStore -> Utxo.UtxoRow -> IO ()
+put hotStore ( Utxo.UtxoRow address txOutRef)  =
```

Comment:
The pair Plutus.Address, TxOutRef is stored in the TVar, as we get them form Marconi.Indexer. @raduom. I could convert the addresses from Plutus.Address  to Cardano Api Address before storing them, however, I chose to store as Plutus Address. When looking the keys up, I convert to Plutus address first and then look them up. My main reason was to avoid an extra conversion as not all keys may be looked up by the users

Reply:
I wouldn't worry too much about that conversion. We have a story which will use cardano-api types instead of plutus types for that indexer.